optimizely/odp/zaius_rest_api_manager.py

```diff
@@ -0,0 +1,86 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+from typing import Optional, List
+
+import requests
+from requests.exceptions import RequestException, ConnectionError, Timeout, JSONDecodeError, InvalidURL
+
+from optimizely import logger as optimizely_logger
+from optimizely.helpers.enums import Errors, OdpRestApiConfig
+from optimizely.odp.odp_event import OdpEvent
+
+"""
+ ODP REST Events API
+ - https://api.zaius.com/v3/events
+ - test ODP public API key = "W4WzcEs-ABgXorzY7h1LCQ"
+
+ [Event Request]
+ curl -i -H 'Content-Type: application/json' -H 'x-api-key: W4WzcEs-ABgXorzY7h1LCQ' -X POST -d '{"type":"fullstack","action":"identified","identifiers":{"vuid": "123","fs_user_id": "abc"},"data":{"idempotence_id":"xyz","source":"swift-sdk"}}' https://api.zaius.com/v3/events
+ [Event Response]
+ {"title":"Accepted","status":202,"timestamp":"2022-06-30T20:59:52.046Z"}
+"""
+
+
+class ZaiusRestApiManager:
+    """Provides an internal service for ODP event REST api access."""
+
+    def __init__(self, logger: Optional[optimizely_logger.Logger] = None):
+        self.logger = logger or optimizely_logger.NoOpLogger()
+
+    def sendOdpEvents(self, api_key: str, api_host: str, events: List[OdpEvent]) -> Optional[bool]:
+        """
+        Dispatch the event being represented by the OdpEvent object.
+
+        Args:
+          api_key: public api key
+          api_host: domain url of the host
+          events: list of odp events to be sent to optimizely's odp platform.
+
+        Returns:
+            retry is True - if network or server error (5xx), otherwise False
+        """
+        can_retry: bool = True
+        url = f'{api_host}/v3/events'
+        request_headers = {'content-type': 'application/json', 'x-api-key': api_key}
+
+        try:
+            response = requests.post(url=url,
+                                     headers=request_headers,
+                                     data=json.dumps(events),
+                                     timeout=OdpRestApiConfig.REQUEST_TIMEOUT)
+
+            response.raise_for_status()
+            can_retry = False
+
+        except (ConnectionError, Timeout):
+            self.logger.error(Errors.ODP_EVENT_FAILED.format('network error'))
+            # we do retry, can_retry = True
+        except JSONDecodeError:
+            self.logger.error(Errors.ODP_EVENT_FAILED.format('JSON decode error'))
+            can_retry = False
+        except InvalidURL:
+            self.logger.error(Errors.ODP_EVENT_FAILED.format('invalid URL'))
+            can_retry = False
+        except RequestException as err:
+            if 400 <= err.response.status_code < 500:
+                self.logger.error(Errors.ODP_EVENT_FAILED.format(err))
```

Comment:
You should log the body of the response here and fall back to the exception if there is no body.

Reply:
@andrewleap-optimizely Do you mean the response body from the request (not from the exc)?
What were you thinking, that there can be a body that is not an exception?

I tested with a few 4xx errors and logged response body from the try block and none of then had any content.
Only "err" - the exception itself seems to contain the error message.

Reply:
I wasn't mocking the response content. Worked when I added it.